Userland/Libraries/LibGfx/AntiAliasingPainter.cpp

```diff
@@ -18,148 +18,131 @@
-                { round(min(mapped_from.x(), mapped_to.x())),
-                    mapped_from.y() - drawing_edge_offset,
-                    round(fabsf(draw_direction.x())),
-                    thickness },
-                color);
-        }
-        return;
+    auto length = mapped_to.distance_from(mapped_from);
+
+    // Axis-aligned lines:
+    if (mapped_from.y() == mapped_to.y()) {
+        auto start_point = (mapped_from.x() < mapped_to.x() ? mapped_from : mapped_to).translated(0, -int_thickness / 2);
+        // FIXME: SVG fill_path() hack:
+        // SVG asks for 1px scanlines at floating point y values, if they're not rounded to a pixel they look faint.
+        start_point.set_y(round_to<int>(start_point.y()));
+        return fill_rect(Gfx::FloatRect(start_point, { length, float(int_thickness) }), color);
+    }
+    if (mapped_from.x() == mapped_to.x()) {
+        auto start_point = (mapped_from.y() < mapped_to.y() ? mapped_from : mapped_to).translated(-int_thickness / 2, 0);
+        return fill_rect(Gfx::FloatRect(start_point, { float(int_thickness), length }), color);
     }
 
-    rotated_rectangle_reference_coords = build_rotated_rectangle(draw_direction, corrected_thickness);
-
-    auto plot = [&](int x, int y, float c) {
-        // ignore rotation if rectangle is fairly small to reduce overhead
-        if (is_straight_line || corrected_thickness < 4) {
-            m_underlying_painter.fill_rect(IntRect::centered_on({ x, y }, size), color.with_alpha(color.alpha() * c));
-        } else if (min(AK::abs(mapped_from.distance_from({ x, y })), AK::abs(mapped_to.distance_from({ x, y }))) >= drawing_edge_offset || AK::abs(mapped_from.distance_from(mapped_to)) < drawing_edge_offset) {
-            // don't draw if we are close to the edge but draw if the whole line is shorter than allowed edge distance
-            m_rotated_rectangle_path.clear();
-            m_rotated_rectangle_path.move_to({ x + rotated_rectangle_reference_coords.p1().x(), y + rotated_rectangle_reference_coords.p1().y() });
-            m_rotated_rectangle_path.line_to({ x + rotated_rectangle_reference_coords.p2().x(), y + rotated_rectangle_reference_coords.p2().y() });
-            m_rotated_rectangle_path.line_to({ x + rotated_rectangle_reference_coords.p3().x(), y + rotated_rectangle_reference_coords.p3().y() });
-            m_rotated_rectangle_path.line_to({ x + rotated_rectangle_reference_coords.p4().x(), y + rotated_rectangle_reference_coords.p4().y() });
-
-            m_rotated_rectangle_path.close();
-            m_underlying_painter.fill_path(m_rotated_rectangle_path, color.with_alpha(color.alpha() * c));
-        }
+    // FIXME: SVG stoke_path() hack:
+    // When painting stokes SVG asks for many thickness * < 1px lines.
+    // It actually wants a thickness * thickness dot centered at that point.
+    if (length < 1.0f)
+        return fill_rect(Gfx::FloatRect::centered_at(mapped_from, { thickness, thickness }), color);
```

Comment:
I still am confused as to why an SVG hack goes into `AntiAliasingPainter::draw_anti_aliased_line`. Can't this workaround live in the SVG code instead, so you can keep this function nice and generic?

Reply:
A little workaround here is just the simplest fix, and I don't think your lose any meaningful generality (since < 1px long lines are quite niche). Could maybe add a flag though `SVGCompatibilityMode::Enable/Disable`?

Reply:
I've added a `draw_line_for_path()` which includes a few hacks for paths. The standard `draw_line()` function now is SVG/Path hack free!